File: src/main/java/fromjavatokotlin/controller/dto/RobotDto.java
package fromjavatokotlin.controller.dto;

import com.fasterxml.jackson.annotation.JsonProperty;
import fromjavatokotlin.model.Identifiable;
import lombok.*;

import javax.validation.constraints.NotBlank;
import java.util.ArrayList;
import java.util.List;

@Getter
@Setter
@ToString(callSuper = true)
@AllArgsConstructor
@NoArgsConstructor
public class RobotDto extends IdentifiableDto {

  @JsonProperty("name")
  private String code;

  private String displayName;

  private Integer mass;

  private Boolean availability;

  private List<AbilityDto> abilities = new ArrayList<>();

  @Data
  @AllArgsConstructor
  @NoArgsConstructor
  public static class AbilityDto {
    @NotBlank
    private String code;
  }
}

File: src/main/kotlin/fromjavatokotlin/controller/dto/RobotDto.kt
package fromjavatokotlin.controller.dto

import com.fasterxml.jackson.annotation.JsonProperty
import java.util.*
import javax.validation.constraints.NotBlank

class RobotDto(
    @JsonProperty("name")
    val code: String? = null,
    val displayName: String? = null,
    val mass: Int? = null,
    val availability: Boolean? = null,
    val abilities: List<AbilityDto> = ArrayList()
) : IdentifiableDto() {

    override fun toString(): String {
        return "RobotDto(super=" + super.toString() + ", code=" + code + ", displayName=" + displayName + ", mass=" + mass + ", availability=" + availability + ", abilities=" + abilities + ")"
    }

    data class AbilityDto(@get:NotBlank val code: String? = null)
}